cardano-api/src/Cardano/Api/TxBody.hs

```diff
@@ -195,65 +226,147 @@ data TxBody era where
        -> TxBody ByronEra
 
      ShelleyTxBody
-       :: Shelley.TxBody StandardShelley
+       :: ShelleyBasedEra era
+       -> Ledger.TxBody (ShelleyLedgerEra era)
        -> Maybe Shelley.MetaData
-       -> TxBody ShelleyEra
-
-deriving instance Eq (TxBody ByronEra)
-deriving instance Show (TxBody ByronEra)
-
-deriving instance Eq (TxBody ShelleyEra)
-deriving instance Show (TxBody ShelleyEra)
+       -> TxBody era
+     -- The 'ShelleyBasedEra' GADT tells us what era we are in.
+     -- The 'ShelleyLedgerEra' type family maps that to the era type from the
+     -- ledger lib. The 'Ledger.TxBody' type family maps that to a specific
+     -- tx body type, which is different for each Shelley-based era.
+
+
+-- The GADT in the ShelleyTxBody case requires a custom instance
+instance Eq (TxBody era) where
+    (==) (ByronTxBody txbodyA)
+         (ByronTxBody txbodyB) = txbodyA == txbodyB
+
+    (==) (ShelleyTxBody era txbodyA txmetadataA)
+         (ShelleyTxBody _   txbodyB txmetadataB) =
+         txmetadataA == txmetadataB
+      && case era of
+           ShelleyBasedEraShelley -> txbodyA == txbodyB
+           ShelleyBasedEraAllegra -> txbodyA == txbodyB
+           ShelleyBasedEraMary    -> txbodyA == txbodyB
```

Comment:
Can this be simplified?

```suggestion
      && txbodyA == txbodyB
```

or is there a reason for doing it this way?

Reply:
No. Can you see why?